### ISSUE TITLE
midnight commander: mouse clicks cause terminal sequences to be emitted in command line

### ISSUE THREAD

**Issue by dl12345** [openwrt/packages#11749]
Currently, if clicking with a mouse anywhere on the mc window, terminal sequences are emitted on the command line. This bug is [reported here](https://midnight-commander.org/attachment/ticket/3954/mc-3954-mouse-with-new-terminfo.patch) and a [patch is available here](https://midnight-commander.org/attachment/ticket/3954/mc-3954-mouse-with-new-terminfo.patch)

The following patch is the patch linked above that has been applied and refreshed with quilt so that it applies cleanly. I confirm that this patch fixes the reported problem. It also makes the mouse work properly in mc and it can now be used to navigate and access the menu

```
commit 472acdb6dd0d30e0642b9134a6fb2c1de4c18f35
Author: Egmont Koblinger <egmont@gmail.com>
Date:   Fri Feb 14 22:14:19 2020 +0100

    Ticket #3954: Fix mouse handling with newer terminfo entries

--- a/lib/tty/key.c
+++ b/lib/tty/key.c
@@ -2123,8 +2123,12 @@ tty_get_event (struct Gpm_Event *event,
 #endif /* KEY_MOUSE */
                           || c == MCKEY_EXTENDED_MOUSE))
     {
-        /* Mouse event */
-        xmouse_get_event (event, c == MCKEY_EXTENDED_MOUSE);
+        /* Mouse event. See tickets 2956 and 3954 for extended mode detection. */
+        xmouse_get_event (event, c == MCKEY_EXTENDED_MOUSE
+#ifdef KEY_MOUSE
+                                 || (c == KEY_MOUSE && xmouse_seq == NULL && xmouse_extended_seq != NULL)
+#endif /* KEY_MOUSE */
+                                 );
         c = (event->type != 0) ? EV_MOUSE : EV_NONE;
     }
     else if (c == MCKEY_BRACKETED_PASTING_START)
--- a/lib/tty/mouse.c
+++ b/lib/tty/mouse.c
@@ -90,8 +90,14 @@ init_mouse (void)
 
     case MOUSE_XTERM_NORMAL_TRACKING:
     case MOUSE_XTERM_BUTTON_EVENT_TRACKING:
-        define_sequence (MCKEY_MOUSE, xmouse_seq, MCKEY_NOACTION);
-        define_sequence (MCKEY_EXTENDED_MOUSE, xmouse_extended_seq, MCKEY_NOACTION);
+        if (xmouse_seq != NULL)
+        {
+            define_sequence (MCKEY_MOUSE, xmouse_seq, MCKEY_NOACTION);
+        }
+        if (xmouse_extended_seq != NULL)
+        {
+            define_sequence (MCKEY_EXTENDED_MOUSE, xmouse_extended_seq, MCKEY_NOACTION);
+        }
         break;
 
     default:
--- a/lib/tty/tty.c
+++ b/lib/tty/tty.c
@@ -310,9 +310,17 @@ tty_init_xterm_support (gboolean is_xter
         }
     }
 
-    /* No termcap for SGR extended mouse (yet), hardcode it for now */
+    /* There's only one termcap entry "kmous", typically containing "\E[M" or "\E[<".
+     * We need the former in xmouse_seq, the latter in xmouse_extended_seq.
+     * See tickets 2956 and 3954 for details. */
     if (xmouse_seq != NULL)
+    {
+        if (strcmp (xmouse_seq, ESC_STR "[<") == 0)
+        {
+            xmouse_seq = NULL;
+        }
         xmouse_extended_seq = ESC_STR "[<";
+    }
 }
 
 /* --------------------------------------------------------------------------------------------- */
```

**Comment by dibdot** [comment]
Are you willing to open a PR for that?

**Comment by dibdot** [comment]
The comments from the author regarding the mentioned ticket doesn't sound like a proper fix (http://midnight-commander.org/ticket/3954), "It's a quick ugly hotfix, no cleanup refactoring." I think we should wait for the next mc release ... at least until the ticket has has been closed with a proper fix.

**Comment by dl12345** [comment]
Fair enough. I see your point. 

This fix does not appear to be scheduled for the [upcoming 4.8.25 release](http://midnight-commander.org/wiki/NEWS-4.8.25).

There could potentially be some libc compatibility issues in an upcoming release, since I note that the above link states a minimal glib compatibility of 2.30, so what effect this will have on musl remains to be seen.

I've applied this patch to my copy of mc in the meantime, since (for me at least) this bug render mc much less usable on OpenWrt. 

At least with this ticket we have the problem identified and the cause documented in case it's reported again.

**Comment by vcunat** [comment]
Fix merged into master upstream: https://github.com/MidnightCommander/mc/commit/d9da121b

**Comment by dibdot** [comment]
should be fixed now in master and 19.07. Thanks @BKPepe
